Replace Legacy Request Data in Messages Admin

tn3rb opened the pull request:
this PR updates the Messages Admin by replacing all uses of the legacy _req_data array with functionality on the new Request class

knazart commented on admin_pages/messages/Messages_Admin_Page.core.php:
```diff
@@ -4400,44 +4356,19 @@ public function save_settings()
         }
 
 
-        if ($this->_req_data['type'] === 'messenger') {
+        if ($type === 'messenger') {
             // this should be an array.
-            $settings  = $this->_req_data['messenger_settings'];
+            $settings  = $this->request->getRequestParam('messenger_settings', [], 'string', true);
             $messenger = $settings['messenger'];
-            // let's setup the settings data
-            foreach ($settings as $key => $value) {
-                switch ($key) {
-                    case 'messenger':
-                        unset($settings['messenger']);
-                        break;
-                    case 'message_types':
-                        unset($settings['message_types']);
-                        break;
-                    default:
-                        $settings[ $key ] = $value;
-                        break;
-                }
-            }
+            // remove messenger and message_types from settings array
+            unset($settings['messenger'], $settings['message_types']);
```
👌

tn3rb replied:
wasn't that just the most bonkers silly code ever?
looping through an array, then using a switch to find a couple of specific keys, so that you can call `unset()` on those keys!
😱

knazart replied:
For sure !

hoseinrafiei replied:
What a code! Thanks for fixing.

knazart reviewed (approved):
👍

Pebblo commented:
This broken messages when you updated a message, you'd get a blank screen with a fatal:

```
[20-Jan-2022 15:36:18 UTC] PHP Fatal error:  Uncaught Error: Call to a member function ID() on null in C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\admin_pages\messages\Messages_Admin_Page.core.php:1481
Stack trace:
#0 C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\core\admin\EE_Admin_Page.core.php(1047): Messages_Admin_Page->_edit_message_template()
#1 C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\core\admin\EE_Admin_Page.core.php(807): EE_Admin_Page->_route_admin_request()
#2 C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\core\admin\EE_Admin_Page_Init.core.php(227): EE_Admin_Page->route_admin_request()
#3 C:\laragon\www\ee4\wp-includes\class-wp-hook.php(303): EE_Admin_Page_Init->initialize_admin_page('')
#4 C:\laragon\www\ee4\wp-includes\class-wp-hook.php(327): WP_Hook->apply_filters('', Array)
#5 C:\laragon\www\ee4\wp-includes\plugin.php(470): WP_Hook->do_action(Array)
#6 C:\laragon\www\ee4\wp-admin\admin.php(259): do_action('event-espresso_...')
#7 {main} thrown in C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\admin_pages\messages\Messages_Admin_Page.core.php on line 1481
```

The commit above fixes that and also an issue with the content being returned as arrays, each field would just show 'Array' and a serialized array was then saved for each field (which isn't the normal).

---

There's still an issue though that I can figure out.

On save all of the HTML content is being stripped from the message template sections:

https://monosnap.com/file/bC1zMOZ33aIxojdusZdSlBgUOBxKL5

To reproduce edit a message type, I used Payment Received and the Event Admin context.

You'll need to set a valid value in the 'To' field and also in the 'Send test email to field', then just save.

garthkoyle commented:
Confirmed what Tony says.

tn3rb commented:
pushed a fix for the HTML getting stripped

garthkoyle commented:
When I test again using the same steps, the fields show `array`, `array`, etc.

![image](https://user-images.githubusercontent.com/3094547/150570033-a5dc7669-df24-4456-9a89-0179bca2f181.png)

tn3rb commented:
@garthkoyle i was getting that too at one point but have since fixed that.
are you sure you have the latest changes?

tn3rb commented:
oh what the %#$%^$#$  I just re-tested and I'm getting that again too
sry... will fix asap

garthkoyle commented:
After saving the TO field input, the main content area goes blank.

tn3rb commented on admin_pages/messages/Messages_Admin_Page.core.php:
```diff
@@ -16,23 +16,40 @@
+    /**
+     * @var EEM_Message_Template_Group
+     */
+    private $MTG_MODEL;
```
added properties and getters for these models because they were called repeatedly with varying names

tn3rb commented:
I don't know about you guys, but I was having problems with the shortcode selectors.
there was a bug in the form input IDs that I fixed, but I also hated the clunky UI so I fixed that too:

<img width="628" alt="Screenshot 2022-01-21 184520" src="https://user-images.githubusercontent.com/1751030/150621946-a27d47a6-507f-44c7-bf8a-e8e95716b6d6.png">

so now when you select a shortcode to add, it will be greyed out and you won't be able to add it again. 
This prevents accidently adding the same shortcode to an input field. That said, if you WANT to have a shortcode in a field more than once, you can simply copy the shortcode within the text field/textarea and then paste it again.

garthkoyle commented:
The shortcode selector on the sides look differently now, and when you click them, it saves/refreshes the page:
![image](https://user-images.githubusercontent.com/3094547/150625353-f346d180-4e00-4bdb-9c43-036017660da1.png)

tn3rb commented:
your browser is caching the assets. 
try opening your browser's dev tools, then click on settings and make sure that caching is turned off

<img width="233" alt="Screenshot 2022-01-23 130526" src="https://user-images.githubusercontent.com/1751030/150697975-24be75d7-f0e5-48a5-8977-937cfbf5f414.png">

Pebblo commented:
Hey Brent,

When I try to rebased master onto this branch I'm getting a conflict from the RequestSanitizer.

Can you please rebase this branch?

hoseinrafiei reviewed (approved):
The codes looks good in my idea.

Pebblo commented:
Messages previews no longer work on this branch.

To reproduce go to Event Espresso -> Messages -> Default Message Templates.

Edit the Payment Received message template, event admin context. (It doesn't match which message or context, but these are what I'm testing)

Click the Preview button.

You'll get an error showing something like:

>There is no preview for this message template available because there are no events created yet. The preview system uses actual events for generating the preview. Go see your events!

(My test site has events and I'm assuming yours will :) )

I've seen an issue with this before and its from the EVT_ID being passed on the query string, the default is currently `0`

Previously that would be swapped out for `false` so that anything that generated a link then skipped it and IIRC this is passed around a fair bit.

For example, the Preview button, the context switcher 'back to edit' button from the preview.

---

You can see this works if you change EVT_ID to a valid event ID and test the links from there.

Or, edit an event and go to the notifications section. Click on the 'Edit' button/link for the payment received message type and then do preview (same as setting the EVT_ID value to that events ID).

tn3rb commented:
ya it turns out it's the WP `add_query_arg()` function:

```php
	foreach ( $qs as $k => $v ) {
		if ( false === $v ) {
			unset( $qs[ $k ] );
		}
	}
```

took me awhile to find that because I was under the impression that the offending logic was within our code somewhere 😆 
so I was looking through file after file after file trying to find it 😕 

that said I fixed this in two places so that:
- we don't pass event IDs of `0` to `EE_Admin_Page::add_query_args_and_nonce()` 
- event IDs of `0` are filtered out of the data that's used to populate the preview

formatted and cleaned up a bunch of other messages files while i was at it

Pebblo commented:
Ok, so I'm getting a few errors on this branch.

Creating custom messages templates is broken.

Go to Event Espresso -> Messages -> Default Message Templates.

Just click 'Create custom' on any message type in that list (same issue with creating custom templates from custom templates)

This error is thrown:

`
[01-Feb-2022 11:12:02 UTC] PHP Fatal error:  Uncaught ArgumentCountError: Too few arguments to function Messages_Admin_Page::_add_message_template(), 0 passed in C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\core\admin\EE_Admin_Page.core.php on line 1047 and exactly 3 expected in C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\admin_pages\messages\Messages_Admin_Page.core.php:1351
Stack trace:
#0 C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\core\admin\EE_Admin_Page.core.php(1047): Messages_Admin_Page->_add_message_template()
#1 C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\core\admin\EE_Admin_Page.core.php(807): EE_Admin_Page->_route_admin_request()
#2 C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\core\admin\EE_Admin_Page.core.php(629): EE_Admin_Page->route_admin_request()
#3 C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\core\admin\EE_Admin_Page.core.php(235): EE_Admin_Page->_page_setup()
#4 C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\admin_p in C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\admin_pages\messages\Messages_Admin_Page.core.php on line 1351
`

With the waitlist add-on enabled there is a fatal thrown when loading the default message templates list;

`
[01-Feb-2022 11:16:59 UTC] PHP Fatal error:  Uncaught ArgumentCountError: Too few arguments to function Messages_Admin_Page::_add_message_template(), 0 passed in C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\core\admin\EE_Admin_Page.core.php on line 1047 and exactly 3 expected in C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\admin_pages\messages\Messages_Admin_Page.core.php:1351
Stack trace:
#0 C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\core\admin\EE_Admin_Page.core.php(1047): Messages_Admin_Page->_add_message_template()
#1 C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\core\admin\EE_Admin_Page.core.php(807): EE_Admin_Page->_route_admin_request()
#2 C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\core\admin\EE_Admin_Page.core.php(629): EE_Admin_Page->route_admin_request()
#3 C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\core\admin\EE_Admin_Page.core.php(235): EE_Admin_Page->_page_setup()
#4 C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\admin_p in C:\laragon\www\ee4\wp-content\plugins\event-espresso-core-reg\admin_pages\messages\Messages_Admin_Page.core.php on line 1351
`

That's being thrown on the "Registration Promoted From Wait List Notification" message type.

(Just for some additional info for the above error, I also have also tested the AUEN add-on on this site and that works. I did that to rule out the error just being from an add-on that adds custom message templates into core)

---

Sending Test emails is broken. With the branch as is, test emails simply don't trigger at all and the message editor shows a message to show the template has been edited. That was due to how the 'test_send' button is used with this setup (it's a flag, if present send the test email and it not, don't) so I pushed a fix for that check but then it show the data isn't set up correctly:

https://monosnap.com/file/OCSZUybefCPAPYAOkwdXz4WHPWOMqE

(That's with my last commit to fix the conditional, just try sending a test email)

Pebblo commented:
I should add, I'm not running into issues with the shortcode selectors like Garth mentioned about, they work fine for me.

tn3rb commented:
@Pebblo 
fixed the fatal error when creating custom messages and the test sends 
but could not recreate the wait list related issue. 

Which branch were you on for the wait list? I pulled the latest master but do not ever receive an error.

tn3rb commented:
thnx @hoseinrafiei, good work.

Can you plz do one more thing now? Switch your local repo to master branch and then retest those few issues that you just discovered and explained above. I want to know if those are existing issues or are a side effect of the work done in this branch.

Pebblo commented:
> 1- Message Template is disabled but shows as active.
> To reproduce, click on one of the grayed out links in Default Message Templates tab and you will see the toggle field is enabled which means it's active.
> 
> https://monosnap.com/file/WU6YmgccyXyaUaY83HWwPcuegGj4Gk
> https://monosnap.com/file/N2PWhewefwI9FsSDnzCNN3rMEsqEYa

We have a hybrid system currently so what you are seeing there is likely expected.

For that message template what is in the 'To' field? I'm guessing it's empty as the Primary Registrant field would be empty by default.

To explain, a while back we didn't have the toggle switches, to enable/disable a message type you had to set or empty the to field in the message. Then the toggle switch was added as a quicker way but we still need to honour that system for past users, so we currently have both.

That means the toggle switch can be ENABLED but if the TO field is empty, the message type is still disabled, that's expected as is.
